src/Data/Singletons/Util.hs

```diff
@@ -251,10 +252,130 @@ ravel :: [DType] -> DType -> DType
 ravel []    res  = res
 ravel (h:t) res = DAppT (DAppT DArrowT h) (ravel t res)
 
+-- Decompose a vanilla function type into its type variables, its context, its
+-- argument types, and its result type. (See
+-- Note [Vanilla-type validity checking during promotion] in
+-- Data.Singletons.Promote.Type for what "vanilla" means.)
+-- If a non-vanilla construct is encountered while decomposing the function
+-- type, an error is thrown.
+--
+-- This should be contrasted with the 'unravelDType' function from
+-- @th-desugar@, which supports the full gamut of function types. @singletons@
+-- only supports a subset of these types, which is why this function is used
+-- to decompose them instead.
+unravelVanillaDType :: DType -> ([DTyVarBndr], DCxt, [DType], DType)
```

Comment:
Should this and the next function be combined? I suppose that would be less performant, but the duplication of code is annoying. Perhaps laziness means that the performance hit would be minimal.

Reply:
That's why I was worried about performance. Of course, `checkVanillaDType` could just call `isJust . unravelVanillaDType`...

Reply:
Actually, I think I've come around to the idea of combining these two functions, since they're just _so_ darn similar. And I don't think the performance hit would be that noticeable.

I've pushed a commit which implements this merger. Let me know if you like the way it looks—I did some refactoring along the way to make it look somewhat nicer.

Reply:
Ah yes, much better now. Thanks!